src/server/mod.rs

```diff
@@ -1,50 +1,101 @@
+mod handler;
+mod https;
+mod service;
+
 use anyhow::Error;
 use hyper::service::{make_service_fn, service_fn};
+use std::net::SocketAddr;
 
+use crate::config::tls::TlsConfig;
 use crate::config::Config;
 
-mod handler;
-mod service;
-
+#[derive(Debug, Clone)]
 pub struct Server {
     config: Config,
 }
 
 impl Server {
-    pub async fn serve(&self) {
+    pub fn new(config: Config) -> Server {
+        Server { config }
+    }
+
+    pub async fn run(self) {
         let address = self.config.address();
         let handler = handler::Handler::from(self.config.clone());
+        let mut server_instances: Vec<tokio::task::JoinHandle<()>> = Vec::new();
+
+        if self.config.tls().is_some() {
+            let handler = handler.clone();
+            let https_config = self.config.tls().unwrap();
+            let host = self.config.address().ip();
+            let port = self.config.address().port() + 1;
+            let address = SocketAddr::new(host, port);
+            let server = self.clone();
+            let task = tokio::spawn(async move {
+                server.serve_https(address, handler, https_config).await;
+            });
+
+            server_instances.push(task);
+        }
+
+        let task = tokio::spawn(async move {
+            self.serve(address, handler).await;
+        });
 
-        let main_svc = make_service_fn(move |_| {
+        server_instances.push(task);
+
+        futures::future::join_all(server_instances).await;
```

Comment:
You don't need to `join_all` here, just wait each task in sequence. The effect is the same, since Tokio tasks run without being polled. As a general rule you should never `join_all` `JoinHandle`s for both efficiency and simplicity reasons.

Reply:
Ohh! That's news to me! Thanks for this tip!